src/Features/Core/Portable/UseCompoundAssignment/AbstractUseCompoundAssignmentDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,207 @@
+        /// operator type (like PlusEqualsToken).
+        /// </summary>
+        private readonly ImmutableDictionary<TSyntaxKind, TSyntaxKind> _assignmentToTokenMap;
+
+        protected AbstractUseCompoundAssignmentDiagnosticAnalyzer(
+            ISyntaxFactsService syntaxFacts,
+            ImmutableArray<(TSyntaxKind exprKind, TSyntaxKind assignmentKind, TSyntaxKind tokenKind)> kinds)
+            : base(IDEDiagnosticIds.UseCompoundAssignmentDiagnosticId,
+                   new LocalizableResourceString(
+                       nameof(FeaturesResources.Use_compound_assignment), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+            _syntaxFacts = syntaxFacts;
+            Utilities.GenerateMaps(kinds, out _binaryToAssignmentMap, out _assignmentToTokenMap);
+        }
+
+        protected abstract TSyntaxKind GetKind(int rawKind);
+        protected abstract TSyntaxKind GetAnalysisKind();
+
+        public override bool OpenFileOnly(Workspace workspace)
+            => false;
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeAssignment, GetAnalysisKind());
+
+        private void AnalyzeAssignment(SyntaxNodeAnalysisContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+            var assignment = (TAssignmentSyntax)context.Node;
+
+            var syntaxTree = assignment.SyntaxTree;
+            var optionSet = context.Options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var option = optionSet.GetOption(CodeStyleOptions.PreferCompoundAssignment, assignment.Language);
+            if (!option.Value)
+            {
+                // Bail immediately if the user has disabled this feature.
+                return;
+            }
+
+            _syntaxFacts.GetPartsOfAssignmentExpressionOrStatement(assignment, 
+                out var assignmentLeft, out var assignmentToken, out var assignmentRight);
+
+            // has to be of the form:  a = b op c
+            // op has to be a form we could convert into op=
+            if (!(assignmentRight is TBinaryExpressionSyntax binaryExpression))
+            {
+                return;
+            }
+
+            var binaryKind = GetKind(binaryExpression.RawKind);
+            if (!_binaryToAssignmentMap.ContainsKey(binaryKind))
+            {
+                return;
+            }
+
+            _syntaxFacts.GetPartsOfBinaryExpression(binaryExpression,
+                out var binaryLeft, out _);
+            
+            // has to be of the form:   expr = expr op ...
+            if (!_syntaxFacts.AreEquivalent(assignmentLeft, binaryLeft))
```

Comment:
this could do _syntaxFacts.AreEquivalent(assignementLeft, binaryRight)?

Reply:
That would be incorrect for many cases.  For exapmle:  `a = 10 - a` is not equivalent to `a -= 10`.

Reply:
sure. but you have an explicit list of operation you support. so that can include explicitly whether commutative is supported or not.

Reply:
I did not want to go this route, because now i'll need to do semantic checks to make sure that the types are only builtins.  For example, non-builtins may not have commutative operators.  They also may not even *define* commutative operators (i.e. tehy may only define one side of hte op, and not the other side).